File: singly_linked_list.c
#include <stdio.h>
#include <stdlib.h>

struct Node {
    int data;
    struct Node* next;
};

struct Node* create_node(int data) {
    struct Node* new_node = (struct Node*)malloc(sizeof(struct Node));
    if (new_node == NULL) {
        printf("Error creating a new node.\n");
        exit(1); // Exiting program if memory allocation fails
    }
    new_node->data = data;
    new_node->next = NULL;
    return new_node;
}

void insert_at_beginning(struct Node** head, int data) {
    struct Node* new_node = create_node(data);
    new_node->next = *head;
    *head = new_node;
}

void insert_at_end(struct Node** head, int data) {
    struct Node* new_node = create_node(data);
    if (*head == NULL) {
        *head = new_node;
        return;
    }
    struct Node* temp = *head;
    while (temp->next != NULL) {
        temp = temp->next;
    }
    temp->next = new_node;
}

void insert_at_between(struct Node* head, int key, int data) {
    struct Node* new_node = create_node(data);
    struct Node* temp = head;
    while (temp != NULL && temp->data != key) {
        temp = temp->next;
    }
    if (temp == NULL) {
        printf("Node with key %d not found.\n", key);
        free(new_node); // Freeing the allocated memory
        return;
    }
    new_node->next = temp->next;
    temp->next = new_node;
}

int delete_node(struct Node** head, int key) {
    struct Node *temp = *head, *prev = NULL;
    if (temp != NULL && temp->data == key) {
        *head = temp->next;
        free(temp);
        return 0;
    }
    while (temp != NULL && temp->data != key) {
        prev = temp;
        temp = temp->next;
    }
    if (temp == NULL) {
        return -1; // Node with given key not found
    }
    prev->next = temp->next;
    free(temp);
    return 0;
}

void print_list(struct Node* head) {
    struct Node* temp = head;
    while (temp != NULL) {
        printf("%d ", temp->data);
        temp = temp->next;
    }
    printf("\n");
}

int main() {
    struct Node* head = NULL;

    insert_at_beginning(&head, 10);
    insert_at_beginning(&head, 20);

    insert_at_end(&head, 30);
    insert_at_end(&head, 40);

    printf("Original List: ");
    print_list(head);

    insert_at_between(head, 20, 16);
    printf("List after inserting 16 after 20: ");
    print_list(head);

    return 0;
}

File: Singly_Linked_List.java
public class singly_linked_list {
        static class Node {
            int data;
            Node next;
    
            Node(int data) {
                this.data = data;
                this.next = null;
            }
        }
    
        static Node createNode(int data) {
            return new Node(data);
        }
    
        static void insertAtBeginning(Node[] head, int data) {
            Node newNode = createNode(data);
            newNode.next = head[0];
            head[0] = newNode;
        }
    
        static void insertAtEnd(Node[] head, int data) {
            Node newNode = createNode(data);
            if (head[0] == null) {
                head[0] = newNode;
                return;
            }
            Node temp = head[0];
            while (temp.next != null) {
                temp = temp.next;
            }
            temp.next = newNode;
        }
    
        static void insertAtBetween(Node head, int key, int data) {
            Node newNode = createNode(data);
            Node temp = head;
            while (temp != null && temp.data != key) {
                temp = temp.next;
            }
            if (temp == null) {
                System.out.println("Node with key " + key + " not found.");
                return;
            }
            newNode.next = temp.next;
            temp.next = newNode;
        }
    
        static int deleteNode(Node[] head, int key) {
            Node temp = head[0];
            Node prev = null;
            if (temp != null && temp.data == key) {
                head[0] = temp.next;
                return 0;
            }
            while (temp != null && temp.data != key) {
                prev = temp;
                temp = temp.next;
            }
            if (temp == null) {
                return -1; // Node with given key not found
            }
            prev.next = temp.next;
            return 0;
        }
    
        static void printList(Node head) {
            Node temp = head;
            while (temp != null) {
                System.out.print(temp.data + " ");
                temp = temp.next;
            }
            System.out.println();
        }
    }
    
        public static void main(String[] args) {
            Node[] head = new Node[1];
            head[0] = null;
    
            insertAtBeginning(head, 10);
            insertAtBeginning(head, 20);
    
            insertAtEnd(head, 30);
            insertAtEnd(head, 40);
    
            System.out.print("Original List: ");
            printList(head[0]);
    
            insertAtBetween(head[0], 20, 25);
            System.out.print("List after inserting 25 after 20: ");
            printList(head[0]);
        }
    }
    
    

